### ISSUE TITLE
Waiting on events other than `inactive` non-reactively

### ISSUE THREAD

**Issue by SteVwonder** [flux-framework/flux-core#2652]
If you are willing to use the flux reactor and write event-driven callback code, you can subscribe to every job event as well as wait for a job to enter the inactive state (either a single, specific job, or the next job from a list).

If you want to program non-reactively (imperatively?), you are currently limited to waiting on a job to enter the inactive state (again, either a single, specific job, or the next job from a list).  It would be great if you could wait for a particular job to experience a particular event (e.g., `alloc`).

One design that came out of a coffee time discussion was to put a wrapper around the `flux_job_event_watch` interfaces.  Something like `flux.job.wait_for_event(jobid, 'alloc')`, which would block until the `alloc` event occurs for `jobid` or until an exception occurs.

**Comment by garlick** [comment]
Not saying we shouldn't do it, but the thing that's bugging me about the proposed  synchronous `wait_for_event` interface is that, non-reactively, it can only block on the events of one job at a time, and that doesn't seem useful in the workflow context. 

Could we provide a high level python class that provides a non-reactive interface to submitting and waiting for jobs all together?  When you instantiate it, you could tell it what states you care about, and it could subscribe to `job-state` events.  For each job you submit via the class, it would internally cache a future.  After submitting some jobs, you could call a blocking "wait" method that would enter the reactor and process both submit responses `job-state` events as they occur.  

Wait could return a (jobid, state) tuple or throw an exception e.g. if submit fails.  It could return `None` when no more notifications are possible for the requested states.  You could alternately call submit(s) and wait(s) to implement a workflow logic without exposing the complexity of the flux reactor.

Maybe it could fetch a job's eventlog on completion and then tack on a result to the tuple for the INACTIVE state transition (jobid, state, success, reason).

Maybe it could use the lightweight `job-manager.wait` RPC if you tell it up front that you only care about  DEPEND (implied by submit response) and INACTIVE states.

Maybe it could use the heavyweight, per-job eventlog watch if you tell it you want even more detail about what's going on (or perhaps if an alternate waitevent method is called).

**Comment by grondo** [comment]
A `workflow` like class for Python is a great idea! I like how it would hide the implementation and actual Flux api calls under the covers. We should check out other  simple workflow or multiprocess management systems in Python to get ideas for the design.

That being said, I think a `wait_for_event`, even though it will only wait on a single job, will still be very useful. As an example, just look at how much we use `flux job wait-event` in the testsuite to synchronize with newly submitted jobs.

**Comment by garlick** [comment]
Fair point!  I'll open a separate issue on the Workflow class idea so this can stay focused on eventlog consumption.
